tinygrad/shape/shapetracker.py

```diff
@@ -27,52 +27,65 @@ def is_contiguous(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> bool: retur
 def filter_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> Tuple[int, ...]:
   return tuple(stride if shp != 1 else 0 for stride, shp in zip(strides, shape))
 
-class View:
-  __slots__ = "shape", "strides", "offset", "mask", "shape_strides", "contiguous"
-  def __init__(self, shape:Tuple[int, ...], strides:Tuple[int, ...], offset:int=0, mask:Optional[Tuple[Tuple[int, int], ...]]=None):
-    self.shape, self.offset = shape, offset
-    self.strides = filter_strides(shape, strides)
-    self.mask = mask
-    self.shape_strides = to_shape_strides(shape, self.strides)
-    self.contiguous: bool = offset == 0 and is_contiguous(shape, self.strides) and mask is None
+class ViewInternal(NamedTuple):
+  shape:Tuple[int, ...]
+  strides:Tuple[int, ...]
+  offset:int
+  mask:Optional[Tuple[Tuple[int, int]]]
+  contiguous:bool
+  shape_strides:Tuple[Tuple[int, int], ...]
 
-  def __repr__(self): return f"View({self.shape}, {self.strides}, {self.offset}, {self.mask})"
-  def key(self): return (self.shape, self.strides, self.offset, self.mask)
-
-  def expr_node_mask(self, idx, valid=None) -> Node:
+@functools.lru_cache(maxsize=None)
+class View(ViewInternal):
+  def __new__(cls, shape, strides=None, offset=0, mask=None):
+    strides_from_shape = strides_for_shape(shape) 
+    strides = strides_from_shape if not strides else filter_strides(shape, strides)
+    contiguous = offset == 0 and is_contiguous(shape, strides) and mask is None
+    return super().__new__(cls, shape, strides, offset, mask, contiguous, to_shape_strides(shape, strides))  
+  def __init__(self, shape, strides=None, offset=0, mask=None, contiguous=False, shape_strides=()): super().__init__()
+
+  @staticmethod
+  @functools.lru_cache(maxsize=None)
+  def expr_node_mask(shape:Tuple[int, ...], mask: Tuple[Tuple[int, int]], idx, valid=None) -> Node:
     expr = [valid] if valid is not None else []
-    if self.mask is not None:
+    if mask is not None:
       acc = 1
-      for ns,(x,y) in reversed(list(zip(self.shape, self.mask))):
+      for ns,(x,y) in reversed(list(zip(shape, mask))):
         base = ((idx//acc) % ns)
         expr += [base >= x, base < y]
         acc *= ns
     return Variable.ands(expr)
 
-  def idxs_to_idx(self, idxs):
-    assert len(idxs) == len(self.shape), "need an idx for all dimensions"
+  @staticmethod
+  @functools.lru_cache(maxsize=None)
+  def idxs_to_idx(shape:Tuple[int, ...], idxs):
+    assert len(idxs) == len(shape), "need an idx for all dimensions"
     acc = 1
     ret = []
-    for tidx,d in reversed(list(zip(idxs, self.shape))):
+    for tidx,d in reversed(list(zip(idxs, shape))):
       ret.append(tidx * acc)
       acc *= d
     return Variable.sum(ret)
 
   # generate an expression if you have a single idx variable
-  def expr_node(self, idx=None) -> Node:
-    if idx is None: idx = Variable('idx', 0, prod(self.shape))
-    ret: List[Node] = [Variable.num(self.offset)]
+  @staticmethod
```

Comment:
I don't understand why this has to change. Why are these staticmethods now? What we wrong with the old way?

Reply:
Speed throught better caching by passing only the required arguments to the functions.

The cache on staticmethods is shared between all instances of this class.

For instance `expr_node_mask`, here the strides or the offset do not matter.

When passing self:
```
Total time: 0.0236809 s
File: /home/rvd/src/roelofvandijk/tinygrad/tinygrad/shape/shapetracker.py
Function: expr_node_mask at line 47

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    47                                             @functools.lru_cache(maxsize=None)
    48                                             @profile
    49                                             def expr_node_mask(self, idx, valid=None) -> Node:
    50      1063        382.6      0.4      1.6      expr = [valid] if valid is not None else []
    51       925        298.4      0.3      1.3      if self.mask is not None:
    52       138         21.3      0.2      0.1        acc = 1
    53       468        359.8      0.8      1.5        for ns,(x,y) in reversed(list(zip(self.shape, self.mask))):
    54       468      13157.6     28.1     55.6          base = ((idx//acc) % ns)
    55       468       5890.3     12.6     24.9          expr += [base >= x, base < y]
    56       468        122.9      0.3      0.5          acc *= ns
    57      1063       3448.0      3.2     14.6      return Variable.ands(expr)
```

As staticmethod:

```
Total time: 0.0167948 s
File: /home/rvd/src/roelofvandijk/tinygrad/tinygrad/shape/shapetracker.py
Function: expr_node_mask at line 47

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    47                                             @staticmethod
    48                                             @functools.lru_cache(maxsize=None)
    49                                             @profile
    50                                             def expr_node_mask(shape:Tuple[int, ...], mask: Tuple[Tuple[int, int]], idx, valid=None) -> Node:
    51       768        312.6      0.4      1.9      expr = [valid] if valid is not None else []
    52       666        136.0      0.2      0.8      if mask is not None:
    53       102         18.3      0.2      0.1        acc = 1
    54       360        296.7      0.8      1.8        for ns,(x,y) in reversed(list(zip(shape, mask))):
    55       360       8986.1     25.0     53.5          base = ((idx//acc) % ns)
    56       360       4423.2     12.3     26.3          expr += [base >= x, base < y]
    57       360        110.0      0.3      0.7          acc *= ns
    58       768       2511.9      3.3     15.0      return Variable.ands(expr)
```

Reply:
Oh, I see. So then they probably shouldn't be on the class and should just be freestanding functions. I didn't understand this before.

Reply:
They absolutely should be freestanding functions.

Reply:
Or functions on the class, not staticmethods. Forgetting perf, what's more readable?